### PR TITLE
Use histogram advice in the Micrometer bridge

### LINKED ISSUE #5292: Cannot get Micrometer custom metrics
**Describe the bug**
I have a Java application instrumented to create custom Micrometer metrics.
The very same Java application uses the OpenTelemetry Java Instrumentation agent to collect metrics using the Prometheus exporter.

I can see Micrometer metrics and OpenTelemetry metrics being exported just fine, but I cannot see Micrometer metrics exported via OpenTelemetry Prometheus exporter.

It's my understanding that from 1.10 the agent should export Micrometer metrics but maybe I get this wrong.

Agent is configured like this:
```
-Dotel.service.name=webportal
-Dotel.exporter.otlp.endpoint=http://tempo.tempo.svc.cluster.local:4317
-Dotel.traces.exporter=otlp
-Dotel.exporter.otlp.compression=gzip
-Dotel.metrics.exporter=prometheus
```

Querying port 9464 produces OTEL metrics like the following:
```
http_client_duration_sum{http_flavor="1.1",http_method="GET",http_status_code="200",http_url="http://localhost:8080/webtng/ediMonitorREST_1_0_0/status",} 301.605403 1643730489470
http_client_duration_bucket{http_flavor="1.1",http_method="GET",http_status_code="200",http_url="http://localhost:8080/webtng/ediMonitorREST_1_0_0/status",le="5.0",} 0.0 1643730489470
http_client_duration_bucket{http_flavor="1.1",http_method="GET",http_status_code="200",http_url="http://localhost:8080/webtng/ediMonitorREST_1_0_0/status",le="10.0",} 0.0 1643730489470
http_client_duration_bucket{http_flavor="1.1",http_method="GET",http_status_code="200",http_url="http://localhost:8080/webtng/ediMonitorREST_1_0_0/status",le="25.0",} 0.0 1643730489470
```

Querying the Mircometer port I get the following:
```
edimonitor_inmessage_incorrect_total{application="edimonitor",edimonitorinstance="0",edimonitorjob="JOB",} 0.0
```
but I cannot find the latter metrics inside the former ones.

**Steps to reproduce**
1. instrument a Java application with custom Java Micrometer metrics
2. auto-instrument a Java application with OTEL agent
3. enable Prometheus exporter for OTEL
4. verify that Micrometer metrics are exported via OTEL agent

**What did you expect to see?**
Metrics from Micrometer reported in the Prometheus exporter endpoint

**What did you see instead?**
Metrics from Micrometer are **not** reported in the Prometheus exporter endpoint

**What version are you using?**
1.10.1

**Environment**
Compiler: (e.g., "AdoptOpenJDK 11.0.6")
OS: (e.g., "Ubuntu 20.04")
Runtime (if different from JDK above): (e.g., "Oracle JRE 8u251")
OS (if different from OS compiled on): (e.g., "Windows Server 2019")

**Additional context**

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
Resolves #5292

I'm marking this as a draft cause this requires a bump of the SDK version to the latest snapshot -- which, unfortunately, introduces a lot of trouble (including changes related to the logs API). We'll want to rebase this PR and merge it just before the 1.26 release.

cc @jack-berg

### CODE REVIEW

File: settings.gradle.kts [hunk 1]
```diff
@@ -22,6 +22,10 @@ plugins {
 dependencyResolutionManagement {
   repositories {
     mavenCentral()
+    maven {
+      name = "sonatype"
+      url = uri("https://oss.sonatype.org/content/repositories/snapshots")
+    }
```
Comment (trask) [inline settings.gradle.kts:28]: did you want to keep this?

File: javaagent/build.gradle.kts [hunk 2]
```diff
@@ -45,7 +48,8 @@ val licenseReportDependencies by configurations.creating {
 
 dependencies {
   bootstrapLibs(project(":instrumentation-api"))
-  // opentelemetry-api is an api dependency of :instrumentation-api, but opentelemetry-api-logs is not
+  // opentelemetry-api is an api dependency of :instrumentation-api, but opentelemetry-api-(events|logs) is not
+  bootstrapLibs("io.opentelemetry:opentelemetry-api-events")
```
Comment (trask) [inline javaagent/build.gradle.kts:52]: is events needed, or just pulled in for completeness (no prob)?

File: instrumentation/micrometer/micrometer-1.5/library/src/main/java/io/opentelemetry/instrumentation/micrometer/v1_5/OpenTelemetryTimer.java [hunk 3]
```diff
@@ -59,12 +61,16 @@ final class OpenTelemetryTimer extends AbstractTimer implements RemovableMeter {
...
         otelMeter
             .histogramBuilder(name)
             .setDescription(Bridging.description(id))
-            .setUnit(TimeUnitHelper.getUnitString(baseTimeUnit))
-            .build();
+            .setUnit(TimeUnitHelper.getUnitString(baseTimeUnit));
+    // micrometer Timers always specify buckets in nanoseconds, we need to convert them to base unit
+    double timeUnitMultiplier = TimeUtils.nanosToUnit(1, baseTimeUnit);
+    setExplicitBucketsIfConfigured(
+        otelHistogramBuilder, distributionStatisticConfig, timeUnitMultiplier);
```
Comment (trask) [inline instrumentation/micrometer/micrometer-1.5/library/src/main/java/io/opentelemetry/instrumentation/micrometer/v1_5/OpenTelemetryTimer.java:72]: maybe pass baseTimeUnit into `setExplicitBucketsIfConfigured()`? I had hard time understanding what `timeUnitMultiplier` meant, and then at least its use would be limited to a single class

**Review by trask — APPROVED** [review]

File: settings.gradle.kts (hunk 1 above)
Comment (mateuszrzeszutek) [reply settings.gradle.kts:28]: Oops 🙈

File: javaagent/build.gradle.kts (hunk 2 above)
Comment (mateuszrzeszutek) [reply javaagent/build.gradle.kts:52]: Not needed, I just noticed that the event API was being shaded and kept in the agent class loader, figured we should probably move it to bootstrap as any other API dep.

File: instrumentation/micrometer/micrometer-1.5/library/src/main/java/io/opentelemetry/instrumentation/micrometer/v1_5/OpenTelemetryTimer.java (hunk 3 above)
Comment (mateuszrzeszutek) [reply instrumentation/micrometer/micrometer-1.5/library/src/main/java/io/opentelemetry/instrumentation/micrometer/v1_5/OpenTelemetryTimer.java]: Good idea, done.

File: javaagent/build.gradle.kts [hunk 4]
```diff
@@ -34,8 +34,11 @@ val javaagentLibs by configurations.creating {
 listOf(baseJavaagentLibs, javaagentLibs).forEach {
   it.run {
     exclude("io.opentelemetry", "opentelemetry-api")
+    exclude("io.opentelemetry", "opentelemetry-api-events")
     exclude("io.opentelemetry", "opentelemetry-api-logs")
     exclude("io.opentelemetry", "opentelemetry-semconv")
+    // metrics advice API
+    exclude("io.opentelemetry", "opentelemetry-extension-incubator")
```
Comment (laurit) [inline javaagent/build.gradle.kts:41]: I think this should also be done in the example distro and agent for testing

**Review by laurit — APPROVED** [review]

File: javaagent/build.gradle.kts (hunk 4 above)
Comment (mateuszrzeszutek) [reply javaagent/build.gradle.kts:41]: I checked that, and both distro and agent-for-testing depend on the javaagent module and build on the `baseJar`/full `shadowJar` configuration; which already contains a properly built bootstrap package.

Comment (laurit) [reply javaagent/build.gradle.kts:41]: So these don't need it?
https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/448b96142a19d8b6b778810fa813ff6bde21cda8/examples/distro/agent/build.gradle#L75-L79

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/448b96142a19d8b6b778810fa813ff6bde21cda8/examples/distro/testing/agent-for-testing/build.gradle#L72-L76

Comment (mateuszrzeszutek) [reply javaagent/build.gradle.kts:41]: Hmm, you may be right after all. Changed 👍

**Pull request merged**